include/openthread/dns.h

```diff
@@ -140,6 +141,93 @@ void otDnsInitTxtEntryIterator(otDnsTxtEntryIterator *aIterator, const uint8_t *
  */
 otError otDnsGetNextTxtEntry(otDnsTxtEntryIterator *aIterator, otDnsTxtEntry *aEntry);
 
+/**
+ * This function is called when a DNS-SD query subscribes a service or service instance.
+ *
+ * @param[in] aContext      A pointer to the application-specific context.
+ * @param[in] aFullName     The null-terminated full service name (e.g. "_ipps._tcp.default.service.arpa."), or
+ *                          full service instance name (e.g. "OpenThread._ipps._tcp.default.service.arpa.").
+ *
+ */
+typedef void (*otDnssdQuerySubscribe)(void *aContext, const char *aFullName);
+
+/**
+ * This function is called when a DNS-SD query unsubscribes a service or service instance.
+ *
+ * @param[in] aContext      A pointer to the application-specific context.
+ * @param[in] aFullName     The null-terminated full service name (e.g. "_ipps._tcp.default.service.arpa."), or
+ *                          full service instance name (e.g. "OpenThread._ipps._tcp.default.service.arpa.").
+ *
+ */
+typedef void (*otDnssdQueryUnsubscribe)(void *aContext, const char *aFullName);
+
+/**
+ * This structure represents information of a discovered service instance for a DNS-SD query.
+ *
+ */
+typedef struct otDnssdServiceInstanceInfo
+{
+    const char *   mFullName;  ///< Full instance name (e.g. "OpenThread._ipps._tcp.default.service.arpa.").
+    const char *   mHostName;  ///< Host name.
+    otIp6Address   mAddress;   ///< Host IPv6 address.
+    uint16_t       mPort;      ///< Service port.
+    uint16_t       mPriority;  ///< Service priority.
+    uint16_t       mWeight;    ///< Service weight.
+    uint16_t       mTxtLength; ///< Service TXT RDATA length.
+    const uint8_t *mTxtData;   ///< Service TXT RDATA.
+    uint32_t       mTtl;       ///< Service TTL (in seconds).
+} otDnssdServiceInstanceInfo;
+
+/**
+ * This structure represents information of a discovered host for a DNS-SD query.
+ *
+ */
+typedef struct otDnssdHostInfo
+{
+    otIp6Address mAddress; ///< Host IPv6 address.
+    uint32_t     mTtl;     ///< Service TTL (in seconds).
+} otDnssdHostInfo;
+
+/**
+ * This function sets DNS-SD server query callbacks. The DNS-SD server calls @p aSubscribe to subscribe to a service or
+ * service instance to resolve a DNS-SD query and @p aUnsubscribe to unsubscribe when the query is resolved or timeout.
+ *
+ * @note @p aSubscribe and @p aUnsubscribe should be both set or unset.
+ *
+ * @param[in] aInstance     The OpenThread instance structure.
+ * @param[in] aContext      A pointer to the application-specific context.
+ * @param[in] aSubscribe    A pointer to the callback function to subscribe a service or service instance.
+ * @param[in] aUnsubscribe  A pointer to the callback function to unsubscribe a service or service instance.
+ *
+ */
+void otDnssdQuerySetCallbacks(otInstance *            aInstance,
+                              void *                  aContext,
```

Comment:
How about moving `aContext` as the last parameter?

A side note here:
In general, unfortunately we don't follow same pattern for callbacks definition. In some callbacks the `aContext` is the first parameter (e.g. in `coap.h` callbacks),  in some others it is last (in `link.h` or `thread.h`). But when registering/setting callback it seems like we put the `aContext` to be always last param.
I have been aiming to harmonize this for a while, but it would be a bigger change and impacts public OT APIs.

Reply:
I don't have strong opinion on it. Putting `aContext` before callback arguments has the advantage that the position of `aContext` won't change when we add more callback functions.